categories/functor.lean

```diff
@@ -0,0 +1,51 @@
+-- Copyright (c) 2017 Scott Morrison. All rights reserved.
+-- Released under Apache 2.0 license as described in the file LICENSE.
+-- Authors: Tim Baumann, Stephen Morgan, Scott Morrison
+
+import .category
+
+open categories
+
+namespace categories.functor
+
+structure {u1 v1 u2 v2} Functor (C : Category.{ u1 v1 }) (D : Category.{ u2 v2 }) :=
+  (onObjects   : C.Obj → D.Obj)
+  (onMorphisms : Π { X Y : C.Obj },
+                C.Hom X Y → D.Hom (onObjects X) (onObjects Y))
+  (identities : ∀ (X : C.Obj),
+    onMorphisms (C.identity X) = D.identity (onObjects X) )
+  (functoriality : ∀ { X Y Z : C.Obj } (f : C.Hom X Y) (g : C.Hom Y Z),
+    onMorphisms (C.compose f g) = D.compose (onMorphisms f) (onMorphisms g) )
```

Comment:
Can this definition be related in any way with the `functor` and `has_map` classes in core lean? In particular, I would like to be able to use `<$>` to denote `onMorphisms`.

Reply:
I’ll have a look.

Personally I think it would be a bad idea to use `<$>`. Perhaps for someone used to Haskell that is helpful, but it’s decidedly arbitrary to most mathematicians. What is the advantage?

Reply:
Over writing `F.onMorphisms`? I thought it would be obvious... Personally I don't care about the particular notation, but I don't think it's possible to get `F` to coerce to two different function types. I would be happy with any random infix operator, as long as I can read the equation by just ignoring it. I think you haven't paid much attention to notation in your repo, and I'd like mathlib equations to be a bit more readable. This is another reason why having `category` be a class is important.